game/src/game/player.rs

```diff
@@ -61,6 +132,563 @@ fn setup_player(q: Query<(&Transform, Entity), Added<PlayerBlueprint>>, mut comm
+//could make derive macro to derive it?
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Idle;
+impl PlayerState for Idle {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Running;
+impl PlayerState for Running {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Falling;
+impl PlayerState for Falling {}
+
+#[derive(Component, Default, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Jumping {
+    //TODO:
+    //can this be frames/ticks instead?
+    airtime: Timer,
+}
+impl Jumping {
+    fn new() -> Self {
+        Jumping {
+            airtime: Timer::from_seconds(0.5, TimerMode::Once),
+        }
+    }
+}
+impl PlayerState for Jumping {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Grounded;
+impl PlayerState for Grounded {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Attacking;
+impl PlayerState for Attacking {}
+
+//=======================================================================
+//
+//TODO: refactor, controller plugin, physics plugin, state plugin
+
+//maybe get rid of idle state?? could jsut play the animation when conditions are met. its a
+//composite of grounded + nothing else...
+fn player_add_idle(
+    query: Query<
+        (Entity, &ActionState<PlayerAction>),
+        (
+            With<PlayerGent>,
+            Without<Idle>,
+            Without<Running>,
+            Without<Jumping>,
+            Without<Falling>,
+            Without<Attacking>,
+        ),
+    >,
+    mut commands: Commands,
+) {
+    for (entity, action_state) in query.iter() {
+        if action_state.get_pressed().len() == 0 {
+            commands.entity(entity).insert(Idle);
+        }
+    }
+}
+
+//can only be idle when grounded and no input...
+fn player_idle(
+    mut query: Query<
+        (
+            Entity,
+            &PlayerGent,
+            &ActionState<PlayerAction>,
+            &Idle,
+        ),
+        With<Grounded>,
+    >,
+    mut commands: Commands,
+) {
+    for (g_ent, _g_marker, action_state, idle) in query.iter() {
+        println!("is idle");
+        // check for direction input
+        let mut direction_vector = Vec2::ZERO;
+        for input_direction in PlayerAction::DIRECTIONS {
+            if action_state.pressed(input_direction) {
+                if let Some(direction) = input_direction.direction() {
+                    direction_vector += Vec2::from(direction);
+                }
+            }
+        }
+        if direction_vector.x != 0.0 {
+            commands.entity(g_ent).insert(idle.new_transition(Running));
+            // } else if direction_vector.x == 0.0 {
+            //     commands.entity(g_ent).insert(Idle);
+        }
+    }
+}
+
+//seprate run and fall/jump movement? y/n?
```

Comment:
My gut feeling is probably no. It is going to get messy and complicated as you start chasing down edge cases in how player movement interacts with the world. In a platformer game, it is very important that the player feels good to control, and that especially matters for controlling the trajectory of jumps. You want to feel confident that you can land your jumps where you want them to. So there is naturally interplay between left/right movement and jumping/falling. If you split them too much in the code (such as by having separate systems to handle each part) you risk not being able to tweak the exact behavior if you need to do that to make the game feel good to control. Having it all in one function actually has benefits in this case. This is one situation where I feel like the general "more granular is better" mindset of ECS programming doesn't apply.